packages/compartment-mapper/src/bundle.js

```diff
@@ -298,6 +341,225 @@ ${''.concat(...modules.map(m => m.bundlerKit.getFunctorCall()))}\
   return bundle;
 };
 
+function wrapFunctorInPrecompiledModule(functorSrc, compartmentName) {
+  const wrappedSrc = `() => (function(){
+  with (this.scopeTerminator) {
+  with (this.globalThis) {
+    return function() {
+      'use strict';
+      return (
+${functorSrc}
+      );
+    };
+  }
+  }
+}).call(getEvalKitForCompartment(${JSON.stringify(compartmentName)}))()`;
+  return wrappedSrc;
+}
+
+// This function is serialized and references variables from its destination scope.
+
+function renderFunctorTable(functorTable) {
+  const entries = Object.entries(functorTable);
+  const lines = entries.map(
+    ([key, value]) => `${JSON.stringify(key)}: ${value}`,
+  );
+  return `{\n${lines.map(line => `  ${line}`).join(',\n')}\n};`;
+}
+
+/**
+ * @param {ReadFn} read
+ * @param {CompartmentMapDescriptor} compartmentMap
+ * @param {Sources} sources
+ * @returns {Promise<string>}
+ */
+export const makeSecureBundleFromAppContainer = async (
+  read,
+  compartmentMap,
+  sources,
+) => {
+  const moduleFunctors = {};
+  const moduleRegistry = {};
+
+  for (const {
+    path,
+    module: { bytes },
+    compartment,
+  } of locationsForSources(sources)) {
+    const textModule = textDecoder.decode(bytes);
+    const moduleData = JSON.parse(textModule);
+    const { __syncModuleProgram__, source, ...otherModuleData } = moduleData;
+    // record module data
+    moduleRegistry[path] = otherModuleData;
+    // record functor
+    if (__syncModuleProgram__) {
+      // esm
+      moduleFunctors[path] = wrapFunctorInPrecompiledModule(
+        __syncModuleProgram__,
+        compartment,
+      );
+    } else {
+      // cjs
+      moduleFunctors[path] = wrapFunctorInPrecompiledModule(
+        source,
+        compartment,
+      );
+    }
+    // other module types?
+  }
+
+  const bundleRuntimeLocation = new URL(
+    './bundle-runtime.js',
+    import.meta.url,
+  ).toString();
+  const runtimeBundle = evadeImportExpressionTest(
+    await makeBundle(read, bundleRuntimeLocation),
```

Comment:
This will not work - read can be anything. It only has access to the bundle-runtime.js if it's coming from node-powers.js

Temporary fix:
```suggestion
      await makeBundle(makeReadPowers({ fs }).read, bundleRuntimeLocation),
```

But I think ultimately it'd be best if we could read the runtime from the archive (backward compatibility wins?)

Reply:
> This will not work - read can be anything. It only has access to the bundle-runtime.js if it's coming from node-powers.js
yeah you're right, we're expecting this to read our internal file

> Alternatively - importing the whole pre-bundled runtime as a function and running a toString on it might work when concatenated with the right invocation.

thats interesting, not sure i understand. are you suggesting pulling it out of the functor table?

Reply:
fixed